### ISSUE TITLE
Current Doxygen master with endless-loop while preprocessing a special cpp file

### ISSUE THREAD

**Issue by jschleus** [doxygen/doxygen#6897]
I just found that the current Doxygen master code seems to go in an endless-loop while preprocessing the file `AnchoredJointNode.cpp` of the software archive `wdune_0.99pl1456.orig.tar.gz` downloaded from https://wdune.ourproject.org/
(at least that suggest the last output line before "hanging").

I used the native Doxyfile generated by "doxygen -g" and only disabled LATEX and set
```
RECURSIVE = YES
```

Remark 1: The current release Doxygen 1.8.15 doesn't have that problem and probably also the master code from 2019-03-17 13:22 (and before) has worked without that problem.

Remark 2: The current Doxygen master code runs well for some other processed software packages.

Remark 3: The mention file is visible for e.g. via
https://fossies.org/linux/privat/wdune_0.99pl1456.orig.tar.gz/wdune-0.99pl1456.orig/src/AnchoredJointNode.cpp

**Comment by albert-github** [comment]
Can you please attach a, small, self contained example (source+configuration file in a tar or zip) that allows us to reproduce the problem?  Please don't add external links as they might not be persistent.

**Comment by albert-github** [comment]
Quick analysis gives that it looks like an endless loop in the used include files through ´pre.l´ this looks like to be introduced by pull request #6859 for the issue: #6856 Doxygen only expands macro defined in header file once when referred multiple times.

Some debugging has to be don in respect to `g_allIncludes`, `absName` and `alreadyIncluded`.

**Comment by jschleus** [comment]
Sorry, I'm afraid I know too little about doxygen to produce the desired example. But I will try to  ensure that the archive `wdune_0.99pl1456.orig.tar.gz` is downloadable at least for the next month under the URL https://fossies.org/linux/privat/wdune_0.99pl1456.orig.tar.gz .

By the way: Although it may not related I want to mention that I see since the above given date (2019-03-17 13:22) for some - but not for all - doxygen-processed packages a drastical increase of processing time by the factor 5-20. If I found more reliable facts I will issue a separate issue.

**Comment by jschleus** [comment]
And still a small supplement, no concrete analysis, just an observation (but eventually nevertheless helpful):

Running doxygen 1.8.15 with the above described simple configuration on the archive `pdns-auth-4.1.8.tar.gz` (downloadable from the original URL https://github.com/PowerDNS/pdns/archive/auth-4.1.8.tar.gz or at least for a month from the alternative URL https://fossies.org/linux/misc/dns/pdns-auth-4.1.8.tar.gz) takes on my system roughly 10 seconds. But using the Doxygen master code from yesterday morning 08:56  (CET timezone) the run wasn't finished after 50 minutes with nearly 100% CPU usage (it was not really hanging but I interrupted it manually). That means in that special case at least an increase of a factor of 300!

I assume the master code from 5 days ago would take also only roughly 10 seconds (sorry, at the moment I do not know how to check out that code).

**Comment by albert-github** [comment]
I downloaded the 2 mentioned files and attached them to this issue (so they are stored permanently):

- [wdune_0.99pl1456.orig.tar.gz](https://github.com/doxygen/doxygen/files/2999130/wdune_0.99pl1456.orig.tar.gz)
- [pdns-auth-4.1.8.tar.gz](https://github.com/doxygen/doxygen/files/2999132/pdns-auth-4.1.8.tar.gz)

**Comment by albert-github** [comment]
I did some extra tests with wdune and I think the code in pre.l function checkAndOpenFile:
```
      if (g_allIncludes.find(absName)==0)
      {
        g_allIncludes.insert(absName,(void *)0x8);
      }
```
should read:
```
      if (g_allIncludes.find(absName)==0)
      {
        g_allIncludes.insert(absName,(void *)0x8);
      }
      else alreadyIncluded=TRUE;
```
but these are preliminary results and I have to run some extra tests, to see whether the results with pdns are OK as well and also if the fix still works for #6856 / #6859.

**Comment by jschleus** [comment]
Ok, let me know if it would be helpful if I search for further software packages that show the described "hanging" or "slowing" problem.

**Comment by albert-github** [comment]
I don't think it is necessary to specifically search, this problem has been identified properly,also the pdns runs through fast, at least in my opinion, I have to test the #6856 / #6859 but this I have to do at my regular work system (will take a little while) after this I can and will make a proposed patch. It never hurts to check other projects for other artifacts.

**Comment by jschleus** [comment]
Although probably a superfluous comment: I can confirm that the above code change solves also on my site the "hanging" resp. "slowing" problem.

**Comment by albert-github** [comment]
I've been looking into the issue and especially in the consequences for #6856 / #6859.
The current situation is not satisfactory (understatement) and also the old situation was not nice (giving incomplete / incorrect information).
The proposed fix (`else  alreadyIncluded=TRUE;`) solves the hanging problem but reintroduces again the old incomplete / incorrect information problem. 
My research has unfortunately not given a solution (:sob:).

@doxygen 
- do you have an idea how to solve this problem?
- for the time being I think it would be better to have the "incomplete / incorrect information" than the hanging situation and it might be better to revert the change of the pull request #6859.

**Comment by jschleus** [comment]
Unfortunately, at least I cannot help since I'm probably the user who uses Doxygen the most but understands it least of all.

Currently https://fossies.org/dox/ offers Doxygen generated documentation for over 3500 packages. So from my very special perspectivev of a fossies.org admin I have currently two main possibilities

1) Running the current (unpatched) release, but use a monitoring utility that kills doxygen after
   a given period (that may dependent from an averaged experience value that can be be derived
   from previously monitored and logged used time values)
   But hours of used CPU time for a single doxygen run isn't currently acceptable on fossies.org.

2) Running an always unofficially patched version of doxygen (probably the preferred and also the currently used solution).

An alternative would be to keep always two versions of doxygen: In case of a kill one can try to rerun with the patched version so principally both problems are in most cases solved (but a very unsatisfactory solution)..

From the perspective of a normal user resp. maintainer of a single software package it looks a little bit different. But since it isn't reasonable to always build a patched self-compiled doxygen binary it may be useful to introduce a configuration directive that controls that different behavior (or sets as a workaround a loop limit in the relevant piece of code ?).

But I hope that a brain that could create such a great and complex software still finds a solution to this annoying problem ;-)

**Comment by albert-github** [comment]
Working on a revert, see #6903

**Comment by albert-github** [comment]
Reverting code has been integrated in master on github (please don't close the issue as this will be done at the moment of a release).

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.8.16.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
